### PR TITLE
[AArch64][GlobalISel] Look through COPY and G_BITCAST while selecting fcvtl2 (fpext)

### PR DESCRIPTION
**Pull request by dzhidzhoev**
It tackles some regressions introduced in
https://reviews.llvm.org/D144670.

### CODE REVIEW

**Review by aemerson — APPROVED** [review]
LGTM.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -4739,14 +4739,18 @@ defm FCMLT  : SIMDFPCmpTwoVector<0, 1, 0b01110, "fcmlt", AArch64fcmltz>;
...
+
 def : Pat<(v4f32 (int_aarch64_neon_vcvthf2fp (v4i16 V64:$Rn))),
           (FCVTLv4i16 V64:$Rn)>;
 def : Pat<(v4f32 (int_aarch64_neon_vcvthf2fp (extract_subvector (v8i16 V128:$Rn),
                                                                 (i64 4)))),
           (FCVTLv8i16 V128:$Rn)>;
 def : Pat<(v2f64 (any_fpextend (v2f32 V64:$Rn))),
           (FCVTLv2i32 V64:$Rn)>;
-def : Pat<(v2f64 (any_fpextend (v2f32 (extract_high_v4f32 (v4f32 V128:$Rn))))),
+def : PatIgnoreCopies<(v2f64 (any_fpextend (v2f32 (extract_high_v4f32 (v4f32 V128:$Rn))))),
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:4753]: Is this needed? Do you have a test case where it makes a difference?

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 2]
```diff
@@ -6777,10 +6777,21 @@ AArch64InstructionSelector::selectExtractHigh(MachineOperand &Root) const {
   MachineRegisterInfo &MRI =
       Root.getParent()->getParent()->getParent()->getRegInfo();
 
-  MachineInstr *Extract = getDefIgnoringCopies(Root.getReg(), MRI);
-  if (Extract && Extract->getOpcode() == TargetOpcode::G_UNMERGE_VALUES &&
-      Root.getReg() == Extract->getOperand(1).getReg()) {
-    Register ExtReg = Extract->getOperand(2).getReg();
+  auto Extract = getDefSrcRegIgnoringCopies(Root.getReg(), MRI);
+  while (Extract && Extract->MI->getOpcode() == TargetOpcode::G_BITCAST)
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6781]: Does this work for big endian? I'm not sure if big endian works for global isel at the moment, but I would prefer not to introduce a lot of subtle cases that assume bitcast does not alter lane layout.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 3]
```diff
@@ -6777,10 +6777,21 @@ AArch64InstructionSelector::selectExtractHigh(MachineOperand &Root) const {
...
+  if (!Extract)
+    return std::nullopt;
+
+  if (Extract->MI->getOpcode() == TargetOpcode::G_UNMERGE_VALUES) {
+    if (Extract->Reg == Extract->MI->getOperand(1).getReg()) {
+      Register ExtReg = Extract->MI->getOperand(2).getReg();
+      return {{[=](MachineInstrBuilder &MIB) { MIB.addUse(ExtReg); }}};
+    }
+  }
+  if (Extract->MI->getOpcode() == TargetOpcode::G_EXTRACT_VECTOR_ELT) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6793]: Why is every G_EXTRACT_VECTOR_ELT considered a high extract? Should it be checking that the Extract is a v2s64 and is extracting from index is 1?

**Review by davemgreen — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 2 above)
Comment (aemerson) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6781]: Big endian currently doesn't work, we have known issues but you're right we should do a LE check here.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 3 above)
Comment (dzhidzhoev) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6794]: Thank you for the catch!

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 1 above)
Comment (dzhidzhoev) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Sorry, I've mixed it up with another change.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 4]
```diff
@@ -6777,11 +6777,30 @@ AArch64InstructionSelector::selectExtractHigh(MachineOperand &Root) const {
...
+    LLT SrcTy = MRI.getType(Extract->MI->getOperand(1).getReg());
+    auto LaneIdx = getIConstantVRegValWithLookThrough(
+        Extract->MI->getOperand(2).getReg(), MRI);
+    if (LaneIdx &&
+        SrcTy == LLT::vector(ElementCount::getFixed(2), LLT::scalar(64)) &&
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6799]: `SrcTy == LLT::fixed_vector(2, 64)` I think would work.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request closed without merging**